File: examples/nodejs/oauth/scan-token.mjs
#!/usr/bin/env node
// scan-token.mjs — locate cached GitHub Copilot / gh CLI / Anthropic OAuth
// tokens on disk and in the environment. Read-only.
//
// Usage:
//   node scan-token.mjs                 # redacted preview
//   node scan-token.mjs --show          # print full tokens (DANGEROUS)
//   node scan-token.mjs --json          # structured findings, one JSON object per line
//
// See README.md for the OAuth-scope caveat: a Copilot session/CAPI token is
// NOT a GitHub PAT and will not authenticate api.github.com REST calls.

import { readdirSync, readFileSync, statSync } from "node:fs";
import { join, extname } from "node:path";
import { homedir, platform } from "node:os";

const argv = new Set(process.argv.slice(2));
const SHOW = argv.has("--show");
const JSON_OUT = argv.has("--json");

// --- candidate locations (mirrors @github/copilot CLI resolution) ---------
const home = homedir();
const xdgCache = process.env.XDG_CACHE_HOME || join(home, ".cache");
const xdgConfig = process.env.XDG_CONFIG_HOME || join(home, ".config");

const candidates = [];
if (process.env.COPILOT_HOME) candidates.push(process.env.COPILOT_HOME);
candidates.push(join(home, ".copilot"));

if (process.env.COPILOT_CACHE_HOME) candidates.push(process.env.COPILOT_CACHE_HOME);
if (platform() === "darwin") candidates.push(join(home, "Library", "Caches", "copilot"));
else if (platform() === "win32" && process.env.LOCALAPPDATA)
  candidates.push(join(process.env.LOCALAPPDATA, "copilot"));
else candidates.push(join(xdgCache, "copilot"));

if (process.env.ANTHROPIC_CONFIG_DIR) candidates.push(process.env.ANTHROPIC_CONFIG_DIR);
candidates.push(join(xdgConfig, "anthropic"));

candidates.push(join(xdgConfig, "gh"));
candidates.push(join(xdgConfig, "github-copilot"));

const ENV_KEYS = [
  "GITHUB_TOKEN", "GH_TOKEN",
  "GITHUB_COPILOT_GITHUB_TOKEN", "COPILOT_GITHUB_TOKEN",
  "GITHUB_COPILOT_API_TOKEN", "GITHUB_VERIFICATION_TOKEN",
  "ANTHROPIC_API_KEY",
];

const SKIP_EXT = new Set([
  ".db", ".wasm", ".so", ".dylib", ".a", ".zip", ".gz",
  ".png", ".jpg", ".jpeg", ".ico", ".pdf", ".bin",
]);
const MAX_SIZE = 1024 * 1024;

const PAT_RE = /(?:gh[psoura]_[A-Za-z0-9]{20,}|github_pat_[A-Za-z0-9_]{40,})/g;
const TOKEN_KEYS = ["access_token", "oauth_token", "github_token", "token",
                    "api_token", "refresh_token", "copilot_token"];

const findings = [];

function redact(v) {
  if (SHOW) return v;
  if (v.length <= 12) return `[short:${v.length}c]`;
  return `${v.slice(0, 8)}…${v.slice(-4)}`;
}

function record(source, key, value) {
  if (!value) return;
  findings.push({ source, key, preview: redact(value), length: value.length });
}

function* walk(dir) {
  let entries;
  try { entries = readdirSync(dir, { withFileTypes: true }); }
  catch { return; }
  for (const e of entries) {
    const p = join(dir, e.name);
    if (e.isDirectory()) yield* walk(p);
    else if (e.isFile()) yield p;
  }
}

function scanFile(path) {
  if (SKIP_EXT.has(extname(path).toLowerCase())) return;
  let st; try { st = statSync(path); } catch { return; }
  if (st.size > MAX_SIZE || st.size === 0) return;

  let text;
  try { text = readFileSync(path, "utf-8"); } catch { return; }
  // Heuristic binary check: NUL byte in first 4KB
  if (text.slice(0, 4096).includes("\0")) return;

  // 1. Token-shaped substrings
  for (const m of text.matchAll(PAT_RE)) record(path, "pattern", m[0]);

  // 2. JSON keys
  if (path.endsWith(".json") || text.trimStart().startsWith("{")) {
    try {
      const parsed = JSON.parse(text);
      const visit = (obj, prefix = "") => {
        if (!obj || typeof obj !== "object") return;
        for (const [k, v] of Object.entries(obj)) {
          const full = prefix ? `${prefix}.${k}` : k;
          if (typeof v === "string" && TOKEN_KEYS.includes(k) && v.length >= 16) {
            record(path, full, v);
          } else if (v && typeof v === "object") visit(v, full);
        }
      };
      visit(parsed);
    } catch {
      // Fall through to regex
    }
  }
  // Always also do a regex pass for malformed/embedded JSON
  const kvRe = /"(access_token|oauth_token|github_token|token|api_token|refresh_token|copilot_token)"\s*:\s*"([^"]+)"/g;
  for (const m of text.matchAll(kvRe)) record(path, m[1], m[2]);

  // 3. YAML-style (gh CLI hosts.yml)
  const yamlRe = /^(\s*)(oauth_token|github_token|token):\s*([^\s#]+)/gm;
  for (const m of text.matchAll(yamlRe)) {
    if (m[3].length >= 16) record(path, m[2], m[3]);
  }
}

// Env vars
for (const k of ENV_KEYS) {
  if (process.env[k]) record("env", k, process.env[k]);
}

// Files
const seen = new Set();
for (const d of candidates) {
  if (seen.has(d)) continue;
  seen.add(d);
  try { statSync(d); } catch { continue; }
  for (const f of walk(d)) scanFile(f);
}

if (JSON_OUT) {
  for (const r of findings) console.log(JSON.stringify(r));
} else {
  console.log("# scan-token.mjs — read-only token scan (--show to reveal full tokens)");
  console.log("# Locations checked:");
  for (const d of candidates) console.log(`#   ${d}`);
  console.log();
  console.log("SOURCE".padEnd(60), "KEY".padEnd(22), "PREVIEW");
  console.log("------".padEnd(60), "---".padEnd(22), "-------");
  for (const r of findings) {
    console.log(r.source.padEnd(60), r.key.padEnd(22), `${r.preview}  (len=${r.length})`);
  }
  if (findings.length === 0) {
    console.log("(no tokens found — run `copilot` or `gh auth login` to populate caches)");
  }
}

File: examples/nodejs/oauth/scan_token.py
#!/usr/bin/env python3
"""scan_token.py — locate cached GitHub Copilot / gh CLI / Anthropic OAuth
tokens on disk and in the environment. Read-only.

Usage:
    python3 scan_token.py              # redacted preview
    python3 scan_token.py --show       # print full tokens (DANGEROUS)
    python3 scan_token.py --json       # one JSON object per line

See README.md for the OAuth-scope caveat: a Copilot session/CAPI token is NOT
a GitHub PAT and will not authenticate api.github.com REST calls.
"""

import argparse
import json
import os
import re
import sys
from pathlib import Path

PAT_RE = re.compile(
    r"(?:gh[psoura]_[A-Za-z0-9]{20,}|github_pat_[A-Za-z0-9_]{40,})"
)
KV_RE = re.compile(
    r'"(access_token|oauth_token|github_token|token|api_token|refresh_token|copilot_token)"'
    r'\s*:\s*"([^"]+)"'
)
YAML_RE = re.compile(
    r"^\s*(oauth_token|github_token|token):\s*([^\s#]+)", re.MULTILINE
)
TOKEN_KEYS = {"access_token", "oauth_token", "github_token", "token",
              "api_token", "refresh_token", "copilot_token"}
SKIP_EXT = {".db", ".wasm", ".so", ".dylib", ".a", ".zip", ".gz",
            ".png", ".jpg", ".jpeg", ".ico", ".pdf", ".bin"}
MAX_SIZE = 1024 * 1024
ENV_KEYS = [
    "GITHUB_TOKEN", "GH_TOKEN",
    "GITHUB_COPILOT_GITHUB_TOKEN", "COPILOT_GITHUB_TOKEN",
    "GITHUB_COPILOT_API_TOKEN", "GITHUB_VERIFICATION_TOKEN",
    "ANTHROPIC_API_KEY",
]


def candidate_dirs() -> list[Path]:
    home = Path.home()
    xdg_cache = Path(os.environ.get("XDG_CACHE_HOME", home / ".cache"))
    xdg_config = Path(os.environ.get("XDG_CONFIG_HOME", home / ".config"))
    out: list[Path] = []
    if v := os.environ.get("COPILOT_HOME"):
        out.append(Path(v))
    out.append(home / ".copilot")
    if v := os.environ.get("COPILOT_CACHE_HOME"):
        out.append(Path(v))
    if sys.platform == "darwin":
        out.append(home / "Library" / "Caches" / "copilot")
    elif sys.platform == "win32":
        if v := os.environ.get("LOCALAPPDATA"):
            out.append(Path(v) / "copilot")
    else:
        out.append(xdg_cache / "copilot")
    if v := os.environ.get("ANTHROPIC_CONFIG_DIR"):
        out.append(Path(v))
    out.append(xdg_config / "anthropic")
    out.append(xdg_config / "gh")
    out.append(xdg_config / "github-copilot")
    return out


def redact(value: str, show: bool) -> str:
    if show:
        return value
    if len(value) <= 12:
        return f"[short:{len(value)}c]"
    return f"{value[:8]}…{value[-4:]}"


def visit_json(obj, prefix: str, sink):
    if isinstance(obj, dict):
        for k, v in obj.items():
            full = f"{prefix}.{k}" if prefix else k
            if isinstance(v, str) and k in TOKEN_KEYS and len(v) >= 16:
                sink(full, v)
            elif isinstance(v, (dict, list)):
                visit_json(v, full, sink)
    elif isinstance(obj, list):
        for i, v in enumerate(obj):
            visit_json(v, f"{prefix}[{i}]", sink)


def scan_file(path: Path, findings: list, show: bool) -> None:
    if path.suffix.lower() in SKIP_EXT:
        return
    try:
        size = path.stat().st_size
    except OSError:
        return
    if size == 0 or size > MAX_SIZE:
        return
    try:
        text = path.read_text(encoding="utf-8")
    except (OSError, UnicodeDecodeError):
        return
    if "\0" in text[:4096]:
        return

    src = str(path)

    def add(key: str, val: str) -> None:
        findings.append({
            "source": src, "key": key,
            "preview": redact(val, show), "length": len(val),
        })

    for m in PAT_RE.finditer(text):
        add("pattern", m.group(0))

    if path.suffix == ".json" or text.lstrip().startswith("{"):
        try:
            visit_json(json.loads(text), "", add)
        except json.JSONDecodeError:
            pass

    for m in KV_RE.finditer(text):
        add(m.group(1), m.group(2))
    for m in YAML_RE.finditer(text):
        if len(m.group(2)) >= 16:
            add(m.group(1), m.group(2))


def main() -> int:
    ap = argparse.ArgumentParser(description="Scan for cached OAuth tokens.")
    ap.add_argument("--show", action="store_true",
                    help="print full tokens (default: redact)")
    ap.add_argument("--json", action="store_true",
                    help="emit one JSON object per finding")
    args = ap.parse_args()

    findings: list[dict] = []
    for k in ENV_KEYS:
        v = os.environ.get(k)
        if v:
            findings.append({
                "source": "env", "key": k,
                "preview": redact(v, args.show), "length": len(v),
            })

    seen: set[Path] = set()
    dirs = candidate_dirs()
    for d in dirs:
        if d in seen or not d.is_dir():
            continue
        seen.add(d)
        for path in d.rglob("*"):
            if path.is_file():
                scan_file(path, findings, args.show)

    if args.json:
        for r in findings:
            print(json.dumps(r))
        return 0

    print("# scan_token.py — read-only token scan (--show to reveal full tokens)")
    print("# Locations checked:")
    for d in dirs:
        print(f"#   {d}")
    print()
    print(f"{'SOURCE':<60} {'KEY':<22} PREVIEW")
    print(f"{'------':<60} {'---':<22} -------")
    for r in findings:
        print(f"{r['source']:<60} {r['key']:<22} {r['preview']}  (len={r['length']})")
    if not findings:
        print("(no tokens found — run `copilot` or `gh auth login` to populate caches)")
    return 0


if __name__ == "__main__":
    sys.exit(main())
